Make cross-importing faster

beccadax opened the pull request:
We previously computed cross-imports by comparing N transitive imports against N transitive imports. This is wasteful, because at least one of the two modules in a pair has to actually declare a cross-import overlay for us to discover one, and the vast majority of modules don’t declare any.

This commit makes us instead compare N transitive imports against M transitive imports which are known to declare at least one cross-import overlay. Since N is potentailly in the thousands while M is perhaps in the double digits, this should be good for a substantial time savings.

However, this optimization has made a test of another cross-import performance optimization fail—not because we have regressed on that, but because it skips work the test case expects us to perform. I have XFAILed that test for now in case someone needs to merge this while I'm on vacation; if that doesn't happen, I'll fix that problem and merge this when I get back.

Fixes <rdar://problem/59538458>.

beccadax committed:
Make cross-importing faster

We previously computed cross-imports by comparing N transitive imports against N transitive imports. This is wasteful, because at least one of the two modules in a pair has to actually declare a cross-import overlay for us to discover one, and the vast majority of modules don’t declare any.

This commit makes us instead compare N transitive imports against M transitive imports which are known to declare at least one cross-import overlay. Since N is potentailly in the thousands while M is perhaps in the double digits, this should be good for a substantial time savings.

However, this optimization has made a test of another cross-import performance optimization fail—not because we have regressed on that, but because it skips work the test case expects us to perform. I have XFAILed that test for now.

Fixes <rdar://problem/59538458>.

brentdax commented:
@swift-ci please test

brentdax commented on include/swift/AST/Module.h:
```diff
@@ -260,6 +260,11 @@ class ModuleDecl : public DeclContext, public TypeDecl {
 
+  /// Returns true if this module is known to not declare any cross-import
+  /// overlays. This doesn't actually load the files, so \c true may be
+  /// incorrect but \c false is never wrong.
```
Should be `\c false may be incorrect but \c true is never wrong`.

harlanhaskins replied:
Drive-by suggestion: would it help to make this a 2-state enum? NoCrossImportOverlays and MayHaveCrossImportOverlays?

brentdax replied:
I decided that's too heavyweight; instead, I'm flipping the sense of the return value and changing its name to `mightDeclareCrossImportOverlays()`, which accurately conveys the uncertainty.

brentdax commented on test/CrossImport/skip-submodules.swift:
```diff
@@ -4,6 +4,11 @@
+// FIXME: This test does not pass because another optimization is now skipping
+// this work entirely. It needs to be redesigned to use modules which have both
+// submodules *and* cross-import overlays.
+// XFAIL: *
```
If someone does merge this, please file a radar on me and add the radar number here.

brentdax replied:
Deleted this test in favor of a more useful one.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -902,55 +917,71 @@ void NameBinder::crossImport(ModuleDecl *M, UnboundImport &I) {
-  // performance fruit.
-
   auto newImports = crossImportableModules.getArrayRef()
                         .slice(nextModuleToCrossImport);
```
Not part of this PR but it would be better to use `drop_front` instead of `slice` as the latter's semantics are not clear from the name.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -902,55 +917,71 @@ void NameBinder::crossImport(ModuleDecl *M, UnboundImport &I) {
-      if (!canCrossImport(oldImport))
-        continue;
+        llvm::makeArrayRef(crossImportableModules.getArrayRef().data(),
+                           &newImport);
```
We should probably annotate the arguments to indicate which overload is being picked. Replacing the `data()` call with `begin()` would help, as otherwise, one might think one is trying to pick the ptr+size overload.

brentdax replied:
I didn't think the comments would really help here, but I added a comment explaining what we're creating.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -978,6 +1017,11 @@ void NameBinder::findCrossImports(UnboundImport &I,
     unboundImports.emplace_back(declaringImport.module.second->getASTContext(),
                                 I, name, declaringImport, bystandingImport);
 
+    if (llvm::is_contained(oppositeNames, name))
```
I think we should default-initialize `oppositeNames` instead of leaving it uninitialized; otherwise, we might access uninitialized memory here?

brentdax replied:
A `SmallVector` is default-initialized without doing anything special—only primitive types (including pointers) are left uninitialized by default in C++.

varungandhi-apple replied:
Whaaaaat, that is so weird. I would've though leaving larger types uninitialized would be more useful than smaller types because there is more work that can be avoided. Sigh.

beccadax committed:
Provide a more useful cross-import submodule test

Removes an unreliable performance-based submodule skipping test and replaces it with a test that checks for a related functional trait instead.

brentdax commented:
@swift-ci test

brentdax commented:
@swift-ci test windows platform

brentdax commented:
@swift-ci smoke test compiler performance

brentdax commented:
@swift-ci please smoke test linux platform
